File: C/Data-Structures/Tree/BinarySearchTree.c
// Binary Search Tree operations in C

#include <stdio.h>
#include <stdlib.h>

struct node {
  int key;
  struct node *left, *right;
};

// Create a node
struct node *newNode(int item) {
  struct node *temp = (struct node *)malloc(sizeof(struct node));
  temp->key = item;
  temp->left = temp->right = NULL;
  return temp;
}

// Inorder Traversal
void inorder(struct node *root) {
  if (root != NULL) {
    // Traverse left
    inorder(root->left);

    // Traverse root
    printf("%d -> ", root->key);

    // Traverse right
    inorder(root->right);
  }
}

// Postorder Traversal
void Postorder(struct node* root) 
{ 
     if (root == NULL) 
        return; 
  
     // first recur on left subtree 
     Postorder(root->left); 
  
     // then recur on right subtree 
     Postorder(root->right); 
  
     // now deal with the node 
     printf("%d ->", root->key); 
} 

// Preorder Traversal
void Preorder(struct node* root) 
{ 
     if (root == NULL) 
          return; 
  
     /* first print data of node */
     printf("%d ->", root->key);   
  
     /* then recur on left sutree */
     Preorder(root->left);   
  
     /* now recur on right subtree */
     Preorder(root->right); 
}     

// Insert a node
struct node *insert(struct node *node, int key) {
  // Return a new node if the tree is empty
  if (node == NULL) return newNode(key);

  // Traverse to the right place and insert the node
  if (key < node->key)
    node->left = insert(node->left, key);
  else
    node->right = insert(node->right, key);

  return node;
}

// Find the inorder successor
struct node *minValueNode(struct node *node) {
  struct node *current = node;

  // Find the leftmost leaf
  while (current && current->left != NULL)
    current = current->left;

  return current;
}

// Deleting a node
struct node *deleteNode(struct node *root, int key) {
  // Return if the tree is empty
  if (root == NULL) return root;

  // Find the node to be deleted
  if (key < root->key)
    root->left = deleteNode(root->left, key);
  else if (key > root->key)
    root->right = deleteNode(root->right, key);

  else {
    // If the node is with only one child or no child
    if (root->left == NULL) {
      struct node *temp = root->right;
      free(root);
      return temp;
    } else if (root->right == NULL) {
      struct node *temp = root->left;
      free(root);
      return temp;
    }

    // If the node has two children
    struct node *temp = minValueNode(root->right);

    // Place the inorder successor in position of the node to be deleted
    root->key = temp->key;

    // Delete the inorder successor
    root->right = deleteNode(root->right, temp->key);
  }
  return root;
}

// Driver code
int main() {
  struct node *root = NULL;
  root = insert(root, 8);
  root = insert(root, 3);
  root = insert(root, 1);
  root = insert(root, 6);
  root = insert(root, 7);
  root = insert(root, 10);
  root = insert(root, 14);
  root = insert(root, 4);
  printf("\nAfter Insertion: ");
  inorder(root);
  printf("\nAfter deleting 10\n");
  root = deleteNode(root, 10);
  inorder(root);
  printf("\nAfter deleting 6\n");
  root = deleteNode(root, 6);
  inorder(root);
  printf("\nInorder traversal: ");
  inorder(root);
  printf("\nPostorder traversal: ");
  Postorder(root);
  printf("\nPreorder traversal: ");
  Preorder(root);
  
  return 0;

}

/* Output:
After Insertion: 1 -> 3 -> 4 -> 6 -> 7 -> 8 -> 10 -> 14 ->
After deleting 10
1 -> 3 -> 4 -> 6 -> 7 -> 8 -> 14 ->
After deleting 6
1 -> 3 -> 4 -> 7 -> 8 -> 14 ->
Inorder traversal: 1 -> 3 -> 4 -> 7 -> 8 -> 14 ->
Postorder traversal: 1 ->4 ->7 ->3 ->14 ->8 ->
Preorder traversal: 8 ->3 ->1 ->7 ->4 ->14 -> */

File: Java/Data-Structures/Binary search tree/binary_search_tree.java
// binary search tree creation, new node insertion , and node deletion
import java.util.*;

class bstree{
    //node structure
    nodes headp;
    static class nodes{
        nodes left;
        int value;
        nodes right;
    }
    //bst creation
    void bstreec(){
    int size;
    nodes duph=headp;
    Scanner in= new Scanner(System.in);
    System.out.println("enter the no of nodes you want to insert");
    int no=in.nextInt();        //no of  nodes asked 
    for(int i=0;i<no;i++)
    {
        if(headp == null){               //if our root is null first create root node
            nodes obj = new nodes ();
            System.out.println("enter the root node");     
            int value=in.nextInt();
            obj.value=value;
            headp=obj;
             duph=headp;

        }                              // if root is assigned or had been created then create futher
        else{
            duph=headp;
            System.out.println("enter next node");
            int value=in.nextInt();
            nodes obj = new nodes();
            obj.value=value;
            while(duph != null){                                 //traversing the tree from root node till we find the position to insert the given node and inserting it
                if(value < duph.value && duph.left != null){
                    duph=duph.left;
                }
                else if(value >duph.value && duph.right != null){
                    duph=duph.right;
                }
                else if( value <duph.value){
                    duph.left=obj;
                    duph=duph.left;     // now going on my newly node which is inserted
                     duph=duph.left;    // going to left of my newly node(it dosent matter you can go right also)
                }
                else{
                    duph.right=obj;
                    duph=duph.right;   // now going on my newly node which is inserted
                     duph=duph.right;    // going to right of my newly node(it dosent matter you can go left also)
                }
            }

        }
        
    }
   
    }

    //insertion of node
    void insert(int value){      //taking the value
            nodes duph=headp; 
            nodes obj = new nodes();
            obj.value=value;
            while(duph != null){             //travrsing the whol
                if(value < duph.value && duph.left != null){
                    duph=duph.left;
                }
                else if(value >duph.value && duph.right != null){
                    duph=duph.right;
                }
                else if( value <duph.value){
                    duph.left=obj;
                    duph=duph.left;             //same explanation as explained above
                     duph=duph.left;            //same explanation as explained above 
                }
                else{
                    duph.right=obj;
                    duph=duph.right;        
                     duph=duph.right;
                    
                }
            }
    }
   

        // Get minimum element in binary search tree
	public  nodes minimumElement(nodes root) {
		if (root.left == null)
			return root;
		else {
			return minimumElement(root.left);
		}
	}


        public  nodes deleteNode(nodes root, int value) {
		if (root == null)
			return null;
		if (root.value > value) {
			root.left = deleteNode(root.left, value);
		} else if (root.value < value) {
			root.right = deleteNode(root.right, value);
 
		} else {
			// if nodeToBeDeleted have both children
			if (root.left != null && root.right != null) {
				nodes temp = root;
				// Finding minimum element from right
				nodes minNodeForRight = minimumElement(temp.right);
				// Replacing current node with minimum node from right subtree
				root.value = minNodeForRight.value;
				// Deleting minimum node from right now
				root.right = deleteNode(root.right, minNodeForRight.value);
 
			}
			// if nodeToBeDeleted has only left child
			else if (root.left != null) {
				root = root.left;
			}
			// if nodeToBeDeleted has only right child
			else if (root.right != null) {
				root = root.right;
			}
			// if nodeToBeDeleted do not have child (Leaf node)
			else
				root = null;
		}
		return root;
	}

      public void treet(nodes poi){   
        //recursive 
           //tree traversal
            if(poi != null){
                System.out.println(poi.value); 
                treet(poi.left);   
                treet(poi.right);
            }

    }
}
public class binary_search_tree{
    public static void main(String [] arrrr){
        Scanner wow = new Scanner(System.in);
        //creting object 
        bstree obj = new bstree();
        obj.bstreec();
        
        System.out.println("tree :-");
        obj.treet(obj.headp);     //display of tree

        //calling to insert
        System.out.println("enter the value you wanna insert");
        int ni=wow.nextInt();
        obj.insert(ni);

        //calling for deletion of node
        System.out.println("enter the value you wanna delete");
        int val=wow.nextInt();
        obj.deleteNode(obj.headp,val);

        System.out.println("tree afterd:-");
        obj.treet(obj.headp);    //display of tree after deleting the node

    }
}
